### PR TITLE
Add update schema

### LINKED ISSUE #2: Add update schema
As a platform user, I want to update a device schema on the fog.

- [ ] Add event listener to update the device schema
- [ ] Update the device schema on the things service
- [ ] Update update device schema message to the connector
- [ ] Update documentation

### PR DESCRIPTION
**Pull request by netoax**
What does this implement/fix? Explain your changes
---------------------------------------------------
Adds update thing's schema capability.

Does this close any currently open issues?
------------------------------------------
Closes #2 

Any other comments\?
-------------------
Documentation is still missing.

Where has this been tested?
---------------------------

**Operating System/Platform:** macOS 10.14 - Darwin 18.0.0

**Go Version:** go1.12.7 darwin/amd64

### CODE REVIEW

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  6
- pkg/user/interactors/create_user.go  2
- pkg/thing/handler/amqp/msg_handler.go  8
- pkg/thing/delivery/amqp/msg_publisher.go  2
- pkg/thing/interactors/interactor.go  1
- pkg/thing/interactors/update_schema_test.go  3
- pkg/thing/interactors/update_schema.go  5
- pkg/user/interactors/create_user_test.go  4
- pkg/user/controllers/user.go  3
- pkg/user/delivery/http/user_proxy.go  4
- cmd/main.go  8
- pkg/thing/delivery/http/thing_proxy.go  4
- pkg/thing/interactors/register_thing_test.go  5
- pkg/thing/delivery/amqp/connector.go  2
- pkg/network/amqp.go  4
- pkg/entities/err_no_perm.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing_test.go  5
- pkg/user/delivery/http/user_proxy.go  4
- cmd/main.go  8
- pkg/user/interactors/create_user_test.go  4
- pkg/thing/handler/amqp/msg_handler.go  8
- pkg/user/interactors/create_user.go  2
- pkg/thing/delivery/http/thing_proxy.go  5
- pkg/thing/delivery/amqp/msg_publisher.go  2
- pkg/thing/interactors/interactor.go  1
- pkg/user/controllers/user.go  3
- pkg/thing/interactors/update_schema_test.go  3
- pkg/thing/interactors/update_schema.go  8
- pkg/entities/err_no_perm.go  1
- pkg/thing/interactors/register_thing.go  6
- pkg/network/amqp.go  4
- pkg/thing/delivery/amqp/connector.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/update_schema_test.go  3
- pkg/user/interactors/create_user_test.go  4
- pkg/user/interactors/create_token_test.go  3
- pkg/thing/interactors/register_thing_test.go  5
- pkg/thing/interactors/update_schema.go  8
- pkg/user/interactors/create_user.go  2
- pkg/thing/delivery/amqp/connector.go  2
- pkg/user/delivery/http/user_proxy.go  5
- pkg/thing/delivery/http/thing_proxy.go  5
- pkg/thing/handler/amqp/msg_handler.go  7
- pkg/user/interactors/create_token.go  2
- pkg/thing/interactors/register_thing.go  6
- pkg/thing/interactors/interactor.go  1
- pkg/user/controllers/user.go  4
- pkg/thing/delivery/amqp/msg_publisher.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/update_schema_test.go  3
- pkg/thing/interactors/update_schema.go  8
- pkg/thing/handler/amqp/msg_handler.go  3
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/update_schema_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  3
- pkg/thing/interactors/update_schema_test.go  3
- pkg/thing/interactors/update_schema.go  8
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/update_schema_test.go  6
- pkg/thing/interactors/register_thing_test.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/update_schema.go  8
- pkg/thing/handler/amqp/msg_handler.go  3
- pkg/thing/interactors/update_schema_test.go  3
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/register_thing_test.go  1
- pkg/thing/interactors/update_schema_test.go  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

File: pkg/thing/handler/amqp/msg_handler.go [hunk 1]
```diff
@@ -21,17 +23,22 @@ type MsgHandler struct {
 	thingInteractor interactors.Interactor
 }
 
+// UpdateSchemaRequest represents the update schema msg
+type UpdateSchemaRequest struct {
```
Comment (vitbaq) [inline pkg/thing/handler/amqp/msg_handler.go]: This struct can be located at knot-babeltower/pkg/network/msg.go

File: pkg/thing/delivery/amqp/msg_publisher.go [hunk 2]
```diff
@@ -18,9 +19,15 @@ type MsgPublisher struct {
 	amqp   *network.Amqp
 }
 
+// UpdatedSchemaResponse represents the update schema response mapped from use case to the AMQP response
+type UpdatedSchemaResponse struct {
```
Comment (vitbaq) [inline pkg/thing/delivery/amqp/msg_publisher.go]: This struct can be located at knot-babeltower/pkg/network/msg.go

File: pkg/thing/delivery/http/thing_proxy.go [hunk 3]
```diff
@@ -52,49 +53,111 @@ func NewThingProxy(logger logging.Logger, hostname string, port uint16) ThingPro
...
+
+func (err errorForbidden) Error() string {
+	return "Error forbidden"
+}
+
+func (err errorConflict) Error() string {
+	return "Error conflict"
+}
+
+func (p proxy) mapErrorFromStausCode(code int) error {
```
Comment (vitbaq) [inline pkg/thing/delivery/http/thing_proxy.go]: Typo? StatusCode

**Review by vitbaq — CHANGES_REQUESTED** [review]

File: pkg/thing/handler/amqp/msg_handler.go [hunk 4]
```diff
@@ -73,6 +95,16 @@ func (mc *MsgHandler) onMsgReceived(msgChan chan network.InMsg) {
...
 				continue
 			}
+		case "schema.update":
+			err := json.Unmarshal(msg.Body, &updateSchemaReq)
+			if err != nil {
+				mc.logger.Error(err)
+				continue
+			}
+
+			authorizationHeader := msg.Headers["Authorization"]
```
Comment (vitbaq) [inline pkg/thing/handler/amqp/msg_handler.go]: Already done on line 89

File: pkg/thing/handler/amqp/msg_handler.go [hunk 5]
```diff
@@ -103,8 +103,11 @@ func (mc *MsgHandler) onMsgReceived(msgChan chan network.InMsg) {
 			}
 
 			authorizationHeader := msg.Headers["Authorization"]
-			mc.logger.Info("Update schema message received")
-			mc.logger.Info(authorizationHeader, updateSchemaReq)
+			err = mc.thingInteractor.UpdateSchema(authorizationHeader.(string), updateSchemaReq.ID, updateSchemaReq.Schema)
```
Comment (vitbaq) [inline pkg/thing/handler/amqp/msg_handler.go]: Create function handleUpdateSchemaMsg. Like handleRegisterMsg().

**Review by vitbaq — CHANGES_REQUESTED** [review]

File: pkg/thing/handler/amqp/msg_handler.go (hunk 1 above)
Comment (joaoaneto) [reply pkg/thing/handler/amqp/msg_handler.go]: Actually, I was thinking about moving `RegisterRequestMsg` and `RegisterResponseMsg` to where they are used and remove that file. At least if the message is used by two files in different packages, it's better to define them where they are used. What do you think?

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/update_schema.go  8
- pkg/thing/interactors/update_schema_test.go  3
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/register_thing_test.go  1
- pkg/thing/interactors/update_schema_test.go  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

File: pkg/thing/handler/amqp/msg_handler.go (hunk 5 above)
Comment (netoax) [reply pkg/thing/handler/amqp/msg_handler.go]: Done.

File: pkg/thing/handler/amqp/msg_handler.go (hunk 4 above)
Comment (netoax) [reply pkg/thing/handler/amqp/msg_handler.go]: Ack. Done.

File: pkg/thing/delivery/http/thing_proxy.go (hunk 3 above)
Comment (netoax) [reply pkg/thing/delivery/http/thing_proxy.go]: Yep, thanks. Done!

File: pkg/thing/delivery/amqp/msg_publisher.go (hunk 2 above)
Comment (netoax) [reply pkg/thing/delivery/amqp/msg_publisher.go]: Done!

File: pkg/thing/handler/amqp/msg_handler.go (hunk 1 above)
Comment (netoax) [reply pkg/thing/handler/amqp/msg_handler.go]: Done!

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/update_schema_test.go  3
- pkg/thing/interactors/update_schema.go  8
- pkg/thing/handler/amqp/msg_handler.go  2
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/update_schema_test.go  11
- pkg/thing/interactors/register_thing_test.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/update_schema.go  8
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/update_schema_test.go  3
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/register_thing_test.go  1
- pkg/thing/interactors/update_schema_test.go  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/update_schema.go  8
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/update_schema_test.go  3
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/register_thing_test.go  1
- pkg/thing/interactors/update_schema_test.go  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Comment by netoax** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/update_schema_test.go  3
- pkg/thing/interactors/update_schema.go  8
- pkg/thing/handler/amqp/msg_handler.go  2
         

Complexity decreasing per file
==============================
+ pkg/thing/delivery/http/thing_proxy.go  -2
         

Clones added
============
- pkg/thing/interactors/register_thing_test.go  1
- pkg/thing/interactors/update_schema_test.go  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=4660884&bid=15606802)

**Review by vitbaq — APPROVED** [review]

**Pull request merged**